sim/src/main/cc/bridges/TracerVModule.cc

```diff
@@ -0,0 +1,367 @@
+    const unsigned timeout = 10000 + s;
+    bool was_done = false;
+    for (unsigned i = 0; i < timeout; i++) {
+      for (auto *bridge : registry.get_all_bridges()) {
+        bridge->tick();
+      }
+
+      if (peek_poke.is_done()) {
+        was_done = true;
+        break;
+      }
+    }
+
+    if (!was_done) {
+      std::cout << "Hit timeout of " << timeout
+                << " tick loops afer a requested " << s << " steps"
+                << std::endl;
+    }
+
+    return was_done;
+  }
+
+  using expected_t = std::pair<uint64_t, std::vector<uint64_t>>;
+
+  std::vector<expected_t> expected_pair;
+
+  void write_iaddr(const unsigned i, const uint64_t x) {
+    peek_poke.poke(namei(i), x, true);
+
+    // as a simplification, also write a modified version of
+    // the PC value as the instruction
+    peek_poke.poke(nameinsn(i), fold(x), true);
+  }
+
+  void write_valid(const unsigned i, const bool x) {
+    peek_poke.poke(namev(i), x, true);
+  }
+
+  int simulation_run() override {
+    tracerv.init();
+
+    // write the header to our expected test output file
+    tracerv.write_header(expected);
+
+    // Reset the DUT.
+    peek_poke.poke("reset", 1, /*blocking=*/true);
+    step(1, /*blocking=*/true);
+    peek_poke.poke("reset", 0, /*blocking=*/true);
+    step(1, /*blocking=*/true);
+
+    // the value of the first cycle as returned from TracerV
+    const uint64_t cycle_offset = 3;
+
+    // modified as we go
+    uint64_t e_cycle = cycle_offset;
+
+    const unsigned tracerv_width = tracerv.max_core_ipc;
+
+    // load MMIO and capture expected outputs
+    auto load = [&](std::vector<uint64_t> iad, std::vector<bool> bit) {
+      std::vector<uint64_t> valid_i;
+      assert(iad.size() == bit.size());
+      for (unsigned i = 0; i < iad.size(); i++) {
+        // std::cout << "loading " << i << " with " << iad[i] << "," << bit[i]
+        // << std::endl;
+
+        write_iaddr(i, iad[i]);
+        write_valid(i, bit[i]);
+
+        // calculate what TraverV should output, and save it
+        if (bit[i]) {
+          valid_i.emplace_back(iad[i]);
+        }
+      }
+
+      // place instructions onto the vector 7 at a time
+      for (size_t i = 0; i < valid_i.size(); i += 7) {
+        auto last = std::min(valid_i.size(), i + 7);
+        std::vector<uint64_t> chunk =
+            std::vector<uint64_t>(valid_i.begin() + i, valid_i.begin() + last);
+        expected_pair.emplace_back(std::make_pair(e_cycle, chunk));
+      }
+
+      e_cycle++;
+    };
+
+    // loop over tests. choose random valids with a simple pattern of iaddr
+    // load into MMIO, and tick the system
+    for (unsigned test_step = 0; test_step < get_total_trace_tests();
+         test_step++) {
+      const uint64_t pull = rand_next(tracerv_width + 1);
+
+      auto pull_iaddr = get_iaddrs(test_step, tracerv_width);
+      auto pull_bits = get_contiguous(pull, tracerv_width);
+
+      load(pull_iaddr, pull_bits);
+      steps(1);
```

Comment:
If i'm understanding this correctly, the DUT is going to stall every single because you're poking new values right. How hard would it be to support runs of time where the bridge is being driven at full throughput? 

Maybe you could capture the values you're enqueuing here in a really larger FIFO in the Chisel DUT, and then let that drain into tracerV (perhaps with some fuzzing of valid) once you say "go". 

Still this is good start.

Reply:
I think this would be easy to do with some support from you. The question here, is this in the scope of this PR? Should we add this to the scope of the wide-tracerv? PR? lets talk